Possible race condition when deleting an EVC

The end-to-end test `tests/test_e2e_10_mef_eline.py::TestE2EMefEline::test_080_create_and_remove_ten_circuits_ten_times ` eventually fail as shown  below:

```
=================================== FAILURES ===================================
______ TestE2EMefEline.test_080_create_and_remove_ten_circuits_ten_times _______

self = <tests.test_e2e_10_mef_eline.TestE2EMefEline instance at 0x7f689c58fe60>

    def test_080_create_and_remove_ten_circuits_ten_times(self):
        """ Tests the creation and removal of ten circuits many times. """
        for x in range(1, 10):
            evcs = {}
            for i in range(400, 410):
                payload = {
                    "name": "evc_%s" % i,
                    "enabled": True,
                    "dynamic_backup_path": True,
                    "uni_a": {
                        "interface_id": "00:00:00:00:00:00:00:01:1",
                        "tag": {"tag_type": 1, "value": i}
                    },
                    "uni_z": {
                        "interface_id": "00:00:00:00:00:00:00:02:1",
                        "tag": {"tag_type": 1, "value": i}
                    }
                }
                api_url = KYTOS_API + '/mef_eline/v2/evc/'
                response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
                assert response.status_code == 201
                data = response.json()
                assert 'circuit_id' in data
                evcs[i] = data['circuit_id']

            time.sleep(10)

            # make sure the evcs are active and the flows were created
            s1, s2 = self.net.net.get('s1', 's2')
            flows_s1 = s1.dpctl('dump-flows')
            flows_s2 = s2.dpctl('dump-flows')
            for vid in evcs:
                evc_id = evcs[vid]
                api_url = KYTOS_API + '/mef_eline/v2/evc/' + evc_id
                response = requests.get(api_url)
                assert response.status_code == 200
                evc = response.json()
                # should be active
                assert evc["active"] is True
                # search for the vlan id
                assert "dl_vlan=%s" % vid in flows_s1
                assert "dl_vlan=%s" % vid in flows_s2
                # search for the cookie, should have two flows
                assert len(re.findall(evc['id'], flows_s1, flags=re.IGNORECASE)) == 2, "round=%d - should have 2 flows but had: \n%s" % (x, flows_s1)
                assert len(re.findall(evc['id'], flows_s2, flags=re.IGNORECASE)) == 2, "round=%d - should have 2 flows but had: \n%s" % (x, flows_s2)

            # Delete the circuits
            for vid in evcs:
                evc_id = evcs[vid]
                api_url = KYTOS_API + '/mef_eline/v2/evc/' + evc_id
                response = requests.delete(api_url)
                assert response.status_code == 200

            time.sleep(10)

            # make sure the circuits were deleted
            api_url = KYTOS_API + '/mef_eline/v2/evc/'
            response = requests.get(api_url)
            assert response.status_code == 200
            assert response.json() == {}
            flows_s1 = s1.dpctl('dump-flows')
            flows_s2 = s2.dpctl('dump-flows')
>           assert len(flows_s1.split('\r\n ')) == 1, "round=%d - should have only 1 flow but had: \n%s" % (x, flows_s1)
E           AssertionError: round=9 - should have only 1 flow but had:
E              cookie=0xaa475b5b6acf3d4d, duration=10.326s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=407 actions=strip_vlan,mod_vlan_vid:3966,output:"s1-eth3"
E              cookie=0xaa475b5b6acf3d4d, duration=10.307s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth3",dl_vlan=3966 actions=strip_vlan,mod_vlan_vid:407,output:"s1-eth1"
E              cookie=0xab00000000000001, duration=231.662s, table=0, n_packets=150, n_bytes=6300, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
E
E           assert 3 == 1
E            +  where 3 = len([' cookie=0xaa475b5b6acf3d4d, duration=10.326s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=407 actions...31.662s, table=0, n_packets=150, n_bytes=6300, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'])
E            +    where [' cookie=0xaa475b5b6acf3d4d, duration=10.326s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=407 actions...31.662s, table=0, n_packets=150, n_bytes=6300, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'] = <built-in method split of str object at 0x55768dafb9f0>('\r\n ')
E            +      where <built-in method split of str object at 0x55768dafb9f0> = ' cookie=0xaa475b5b6acf3d4d, duration=10.326s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=407 actions=...231.662s, table=0, n_packets=150, n_bytes=6300, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'.split

tests/test_e2e_10_mef_eline.py:708: AssertionError
```

This test was run with the `DEPLOY_EVCS_INTERVAL` adjusted to 5s instead of the default 60s. Also, the same test was repeated 10 times and we had one failure.

When investigating the issue with Kytos running in debug mode, the following log entries seems to point where is the issue (filtering only the important log entries:

```
# grep 407 /var/log/syslog
...
Jan 13 03:16:41 3fdbfe893925 2022-01-13 03:16:41,809 kytos.napps.kytos/mef_eline:INFO (Thread-1145) main:321:  Removing EVC(475b5b6acf3d4d, evc_407)
Jan 13 03:16:41 3fdbfe893925 2022-01-13 03:16:41,854 kytos.napps.kytos/flow_manager:DEBUG (ThreadPoolExecutor-0_19) main:415:  stored_flows_list on switch 00:00:00:00:00:00:00:01 by cookie: 0xaa475b5b6acf3d4d: [{'_id': '28c2a57078efebf08de7205ed50f6c62', 'flow': {'match': {'in_port': 1, 'dl_vlan': 407}, 'cookie': 12269876157986323789, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3995}, {'action_type': 'output', 'port': 3}]}, 'created_at': '2022-01-13T03:16:29', 'state': 'installed'}, {'_id': 'd70be044f7283051e1a413aac81f6ad3', 'flow': {'match': {'in_port': 3, 'dl_vlan': 3995}, 'cookie': 12269876157986323789, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 407}, {'action_type': 'output', 'port': 1}]}, 'created_at': '2022-01-13T03:16:29', 'state': 'installed'}]
Jan 13 03:16:41 3fdbfe893925 2022-01-13 03:16:41,848 kytos.napps.kytos/flow_manager:INFO (Thread-1146) main:725:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12269876157986323789, 'cookie_mask': 18446744073709551615}], 'force': True}
Jan 13 03:16:41 3fdbfe893925 2022-01-13 03:16:41,914 kytos.napps.kytos/flow_manager:INFO (Thread-1147) main:725:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12269876157986323789, 'cookie_mask': 18446744073709551615}], 'force': True}
Jan 13 03:16:41 3fdbfe893925 2022-01-13 03:16:41,971 kytos.napps.kytos/mef_eline:INFO (mef_eline) main:89:  EVC(475b5b6acf3d4d, evc_407) enabled but inactive - redeploy
Jan 13 03:16:41 3fdbfe893925 2022-01-13 03:16:41,992 kytos.napps.kytos/flow_manager:INFO (Thread-1149) main:725:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12269876157986323789, 'cookie_mask': 18446744073709551615}], 'force': True}
Jan 13 03:16:42 3fdbfe893925 2022-01-13 03:16:42,021 kytos.napps.kytos/flow_manager:INFO (Thread-1150) main:725:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12269876157986323789, 'cookie_mask': 18446744073709551615}], 'force': True}
Jan 13 03:16:42 3fdbfe893925 2022-01-13 03:16:42,196 kytos.napps.kytos/mef_eline:INFO (Thread-1145) main:328:  EVC removed. EVC(475b5b6acf3d4d, evc_407)
Jan 13 03:16:42 3fdbfe893925 2022-01-13 03:16:42,304 kytos.napps.kytos/flow_manager:INFO (Thread-1153) main:725:  Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 1, 'dl_vlan': 407}, 'cookie': 12269876157986323789, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 3966}, {'action_type': 'output', 'port': 3}]}, {'match': {'in_port': 3, 'dl_vlan': 3966}, 'cookie': 12269876157986323789, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 407}, {'action_type': 'output', 'port': 1}]}], 'force': False}
...
```

The above log entries basically are showing that the delete operation started running at `2022-01-13 03:16:41,809` and the consistency routine has run for the same EVC starting at `2022-01-13 03:16:41,971`. The first step when deleting the EVC is to remove the current flows (https://github.com/kytos-ng/mef_eline/blob/309f7d18c7f03acb2f82a9df430f231bf591e5dd/main.py#L321) - which is done async - and then deactivate the EVC (https://github.com/kytos-ng/mef_eline/blob/309f7d18c7f03acb2f82a9df430f231bf591e5dd/main.py#L322). Both actions are done without using the `EVC.Lock`, which can lead to the consistency routine being executed while the EVC is being removed. Thus, the shared resource (the EVC) will be inconsistent between the two execution threads.

--- a/main.py
+++ b/main.py
@@ -318,12 +318,13 @@ class Main(KytosNApp):
             raise NotFound(result) from NotFound
 
         log.info("Removing %s", evc)
-        evc.remove_current_flows()
-        evc.deactivate()
-        evc.disable()
-        self.sched.remove(evc)
-        evc.archive()
-        evc.sync()
+        with evc.lock:
+            evc.remove_current_flows()
+            evc.deactivate()
+            evc.disable()
+            self.sched.remove(evc)
+            evc.archive()
+            evc.sync()
         log.info("EVC removed. %s", evc)
         result = {"response": f"Circuit {circuit_id} removed"}
         status = 200

